icu4j/main/classes/core/src/com/ibm/icu/message2/MessageFormatter.java

```diff
@@ -0,0 +1,333 @@
+// © 2022 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+package com.ibm.icu.message2;
+
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * {@code MessageFormatter} is the next iteration of {@link com.ibm.icu.text.MessageFormat}.
+ *
+ * <p>For more than two years the Message Format Working Group (MFWG) worked on
+ * developing an industry standard for the representation of localizable message
+ * strings.</p>
+ *
+ * <p>MFWG will recommend how to remove redundancies and pitfalls, make the syntax
+ * more usable, and support more complex features, such as gender, inflections,
+ * and speech. MFWG will also consider the integration of the new standard with
+ * programming environments, including, but not limited to, ICU, HTML, and
+ * ECMAScript, and with localization platform interchange. The output of MFWG
+ * will be a specification for the new syntax, which is expected to be on track
+ * to become a Unicode Technical Standard.</p>
+ *
+ * <p>The reasoning for this effort is shared in the
+ * <a href="https://github.com/unicode-org/message-format-wg/blob/main/docs/why_mf_next.md">“Why
+ * MessageFormat needs a successor”</a> document.</p>
+ *
+ * <p>The project is hosted on <a href="https://github.com/unicode-org/message-format-wg">GitHub</a>.</p>
+ *
+ * <p>The current specification for the syntax and data model can be found <a href=
+ * "https://github.com/unicode-org/message-format-wg/blob/develop/spec/syntax.md">here</a>.
+ * </p>
+ *
+ * <p>This tech preview implements enough of the MessageFormat functions to be useful,
+ * but the final set of functions and the parameters accepted by those functions is not yet finalized.</p>
+ *
+ * <p>These are the functions interpreted right now:</p>
+ *
+ * <table border="1">
+ * <tr>
+ *   <td rowspan="4">{@code datetime}</td>
+ *   <td>Similar to the ICU {@code "date"} and {@code "time"}.</td>
+ * </tr>
+ *
+ *   <tr><td>{@code datestyle} and {@code timestyle}<br>
+ *   Similar to {@code argStyle : short | medium | long | full}.<br>
+ *   Same values are accepted, but we can use both in one placeholder,
+ *   for example <code>{$due :datetime datestyle=full timestyle=long}</code>.
+ *   </td></tr>
+ *
+ *   <tr><td>{@code pattern}<br>
+ *   Similar to {@code argStyle = argStyleText}.<br>
+ *   This is bad i18n practice, and will probably be dropped.<br>
```

Comment:
Change to: This is included just to support migration to MF2.0.

Reply:
Done.